Std/Data/Array/Match.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2023 F. G. Dorais. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: F. G. Dorais
+-/
+import Std.Data.Array.Basic
+import Std.Data.Array.Lemmas
+
+namespace Array
+
+/-- Prefix table for the Knuth-Morris-Pratt matching algorithm
+
+  This is an array of the form `t = [(x₀,n₀), (x₁,n₁), (x₂, n₂), ...]` where for each `i`, `nᵢ` is
+  the length of the longest proper prefix of `xs = [x₀,x₁,...,xᵢ]` which is also a suffix of `xs`.
+-/
+structure PrefixTable (α : Type _) extends Array (α × Nat) where
+  /-- Validity condition to help with termination proofs -/
+  valid : (h : i < toArray.size) → toArray[i].2 ≤ i
+
+instance : Inhabited (PrefixTable α) where
+  default := ⟨#[], fun .⟩
+
+/-- Returns the size of the prefix table -/
+abbrev PrefixTable.size (t : PrefixTable α) := t.toArray.size
+
+/-- Returns the pattern array of the prefix table -/
+abbrev PrefixTable.pattern (t : PrefixTable α) : Array α := t.toArray.map Prod.fst
```

Comment:
This operation is not O(1), which I think messes up the asymptotic order of the algorithm because it is called in `PrefixTable.step`. Maybe the numbers array should just be a parallel array instead of being interleaved with the pattern?